src/ostree/ot-builtin-prune.c

```diff
@@ -191,6 +191,16 @@ ostree_builtin_prune (int argc, char **argv, OstreeCommandInvocation *invocation
     pruneflags |= OSTREE_REPO_PRUNE_FLAGS_NO_PRUNE;
   if (opt_commit_only)
     pruneflags |= OSTREE_REPO_PRUNE_FLAGS_COMMIT_ONLY;
+  if (opt_refs_only && (opt_retain_branch_depth || opt_keep_younger_than))
+    /* The flag below will tell ostree to look only at refs when gathering
+     * objects. This will speed up operations when trying to truncate branch history.
+     *
+     * We don't want to blindly pass this flag. Consider branch: HEAD->A->B->C,
+     * if we delete commit B with --delete-commit flag, we will not be able to 
+     * retrieve orphaned commit C if we only look at refs. Thus C will not be pruned,
+     * which may be different from what the user expects.
```

Comment:
So I think `opt_refs_only` is really the common case.  And from that PoV, we *are* changing behavior in this way by default.

Now, the fact that you're doing this in the CLI does limit the "blast radius" to some degree.  But I think it may lead to confusion if the CLI has special behavior not directly reflected in the API.

Hmmm.  Thinking through this...are there any cases where one wants `OSTREE_REPO_PRUNE_FLAGS_COMMIT_ONLY` but *not* `OSTREE_REPO_PRUNE_FLAGS_LIST_REFS_ONLY`? 

To me what is special here is `OSTREE_REPO_PRUNE_FLAGS_COMMIT_ONLY` - that's not something we do in a "normal prune" (as used by e.g. the default cleanup after `ostree admin upgrade`).

What if we dropped this public flag and turned on this behavior when `OSTREE_REPO_PRUNE_FLAGS_COMMIT_ONLY` is provided?

Reply:
That's what I was just thinking. I don't understand why a new flag is needed, but I've never really used these modes. It seems to me that `COMMIT_ONLY` should have one simple effect - it filters the objects in both the full set and the wanted set. That allows a big optimization by not having to traverse commits. Really you don't have to traverse anything - you can simply scan the repo for commit objects. Shouldn't that just be the behavior for `COMMIT_ONLY` without any other flags? Am I missing something?

Reply:
> Thinking through this...are there any cases where one wants OSTREE_REPO_PRUNE_FLAGS_COMMIT_ONLY but not OSTREE_REPO_PRUNE_FLAGS_LIST_REFS_ONLY

So currently, `OSTREE_REPO_PRUNE_FLAGS_LIST_REFS_ONLY` gathers all objects by traversing refs (*not* just commit objects, unless `--commit-only` is provided). In the case that a user would want to truncate the history of a branch (`ostree prune --refs-only --keep-younger-than=<DATE>`) , but they don't want to only delete commit objects, it would make sense to only gather objects via traversal of refs.

> What if we dropped this public flag and turned on this behavior when OSTREE_REPO_PRUNE_FLAGS_COMMIT_ONLY is provided?

I'm OK with ignoring the case mentioned above and switching to only supporting this for `--commit-only`. I felt that supporting the new optimization for  `ostree prune --refs-only` instead of only `--commit-only` would help generalise it. If we were to remove `OSTREE_REPO_PRUNE_FLAGS_COMMIT_ONLY` and instead just turn on the functionality when `--commit-only` is passed, we would have to disallow certain combinations of `--commit-only` and other flags. An example would be the case mentioned in the comment referenced above.

Reply:
> It seems to me that COMMIT_ONLY should have one simple effect - it filters the objects in both the full set and the wanted set. That allows a big optimization by not having to traverse commits. Really you don't have to traverse anything - you can simply scan the repo for commit objects. Shouldn't that just be the behavior for COMMIT_ONLY without any other flags? Am I missing something?

Currently, `--commit-only` filters the reachable set such that it only includes commit objects. However, it does not filter the set of all objects. We want the flag to do both by adding this change. At the same time, we want to avoid the cost of using readdir on objects that we don't need (thus why we are traversing refs).  The reason why I added another flag to the prune API is because I want to disallow certain cases ( I.E using the new optimization with `--delete-commit`). If we don't add a flag to turn the optimization off, then we need to disallow the use of `--commit-only` with `--delete-commit` which is a little awkward since we allowed that possibility in the last release.

Sorry if I'm over-explaining here - this is also just to clear my head.

Reply:
OK, I understand that. What I'm saying is that you can also achieve the filtering on the set of all objects by using `list_commits_starting_with` instead of `list_repo_objects`. That can be used without introducing flags or changing the definition of "all objects". However, it may not achieve enough of an optimization since you're still scanning all the objects.